app/celery/process_delivery_status_result_tasks.py

```diff
@@ -1,13 +1,296 @@
+
+    # the race condition scenario if we got the delivery status before we actually record the sms
+    if should_retry or (notification is None):
+        self.retry(queue=QueueNames.RETRY)
+
+    if should_exit:
+        return False
+
+    try:
+        # calculate pricing
+        _calculate_pricing(price_in_millicents_usd, notification, notification_status, number_of_message_parts)
+        current_app.logger.info(
+            "Delivery Status callback return status of %s for notification: %s",
+            notification_status,
+            notification.id,
+        )
+
+        # statsd - metric tracking of # of messages sent
+        _increment_statsd(notification, provider_name, notification_status)
+
+        # check if payload is to be include in cardinal set in the service callback is (service_id, callback_type)
+        if not _get_include_payload_status(self, notification):
+            payload = {}
+        check_and_queue_callback_task(notification, payload)
+        return True
+
+    except Retry:
+        # This block exists to preempt executing the "Exception" logic below.  A better approach is
+        # to catch specific exceptions where they might occur.
+        raise
+    except Exception as e:
+        current_app.logger.exception(e)
+        self.retry(queue=QueueNames.RETRY)
+
     return True
+
+
+def attempt_to_get_notification(
+        reference: str, notification_status: str, event_timestamp_in_ms: str
+) -> Tuple[Notification, bool, bool]:
+    """ Attempt to get the notification object and determine whether the Celery Event should be retry or exit"""
+    should_retry = False
+    notification = None
+    try:
+        notification = dao_get_notification_by_reference(reference)
+        should_exit = check_notification_status(notification, notification_status)
+    except NoResultFound:
+        message_time = datetime.datetime.fromtimestamp(
+            math.floor(float(event_timestamp_in_ms) / 1000)
+        )
+        if datetime.datetime.utcnow() - message_time < datetime.timedelta(minutes=5):
+            current_app.logger.info(
+                "Delivery Status callback event for reference %s was received less than five minutes ago.",
+                reference,
+            )
+            should_retry = True
+        else:
+            current_app.logger.warning(
+                "notification not found for reference: %s (update to %s)",
+                reference,
+                notification_status,
+            )
+        statsd_client.incr("callback.delivery_status.no_notification_found")
+        should_exit = True
+    except MultipleResultsFound:
+        current_app.logger.warning(
+            "multiple notifications found for reference: %s (update to %s)",
+            reference,
+            notification_status,
+        )
+        statsd_client.incr("callback.delivery_status.multiple_notifications_found")
+        should_exit = True
+
+    return notification, should_retry, should_exit
```

Comment:
This can throw a `NameError` if it fails under a different exception. Best to set this at the start.